### DISCUSSION TITLE
Feature Request: Server Screen Window

### DISCUSSION THREAD

**Discussion by AndyMcProducer** [jamulussoftware/jamulus#1029] [UI Related]
I can't seem to set the label from issue to feature request.

Server Screen Window, let me explain what I mean. 
Basically a Extra text field called Server Screen on the server and an extra button by settings and chat buttons called Server Screen.
When clicked it opens a window that loads the page or embed code which is entered in the Server Screen field on the Jamulus
window.

This way servers can have there song book web sites, a jitsu or zoom page or other. Or a plash type page with info about the server, links to band members, there songs and other info.

Reason for button, people can turn it off if they don't want to see it., Backwards compatibility, when omitted it just hits a basic local page saying this server doesn't have a screen page.

I with I still programmed as I'd add this in asap.

**Comment by npostavs** [comment]
Wouldn't a link in the server welcome message accomplish the same?

**Comment by AndyMcProducer** [comment]
No.

**Comment by npostavs** [comment]
Um, care to elaborate on what the difference would be? Perhaps I've misread your proposal.

**Comment by AndyMcProducer** [comment]
Chat window is almost this, but doesn't support iframe or other stuff. I'm guessing it more supports basic html?

**Comment by npostavs** [comment]
> I'm guessing it more supports basic html?

Yes, specifically, it supports links which would open in a browser window.

**Comment by gilgongo** [comment]
I think this may be difficult due to payload size limitations over UDP. At least, that was mentioned when discussing the issue of a free text field for licence notices, to which this seems similar. Perhaps some engineers may be able to confirm.

**Comment by hoffie** [comment]
> I think this may be difficult due to payload size limitations over UDP.

Yes, sending arbitrary large HTML would require multiple packets and would therefore need some sub-protocol for ordering/end-of-message delimiting, as far as I understand.

However, I'm more concerned about the client-side part. While Qt seems to have support for embedding web content, I'm not sure if it's a good idea. Web browsers tend to be security-sensitive. When bundling/statically linking Qt (I assume this is the case on Windows at least?) one would have to carry that burden. I'm also unsure if such an embedded browser would adequately support Jitsi/Zoom etc. as requested as this would require camera support. It might also make it harder for people to get good Jamulus performance (network- and CPU-wise).

I'd vote for implementing this with links in the welcome chat message which can be opened in the user's preferred, properly configured and known browser as @npostavs suggested.

**Comment by pljones** [comment]
Jamulus is not a web browser.  Formatting a welcome message is one thing but full web page support is another.  Any need for anything more extensive should be through the provision of links.

**Comment by AndyMcProducer** [comment]
But it should be possible to bring up a window that is a container of the default browser, this way there is effectively no route into jamulus.
________________________________

**Comment by AndyMcProducer** [comment]
This explains about a Chrome window but I'd have through was a way to just pull up a container window of default browser in system.
It's been done many times before on software and if something doesn't work  ya use to have to make sure flash or another plugin was installed.
But nowadays that's not needed as everything needed is normally there.
https://doc.qt.io/qt-5/topics-web-content.html
Integrating Web Content | Qt 5.15<https://doc.qt.io/qt-5/topics-web-content.html>
©2021 The Qt Company Ltd. Documentation contributions included herein are the copyrights of their respective owners. The documentation provided herein is licensed under the terms of the GNU Free Documentation License version 1.3 as published by the Free Software Foundation.
doc.qt.io

________________________________

**Comment by AndyMcProducer** [comment]
This might help some. This is like a recipe browser but the idea is there to make something basic which could become something more later.
Plus when peeps integrate audio stuff into browsers they really shoudl be more an extension which can run code outside of the browser environment to.
Example the audio engine to handle the asio part.
https://doc.qt.io/qt-5/qtwebengine-webengine-recipebrowser-example.html
WebEngine Recipe Browser | Qt WebEngine 5.15.2<https://doc.qt.io/qt-5/qtwebengine-webengine-recipebrowser-example.html>
Next, we add a RowLayout item so we can divide the window into two parts: one being a custom RecipeList item containing the recipe titles, and the other being the WebEngineView, which shows the recipe details.The spacing is set to zero so the items are positioned directly next to each other.
doc.qt.io

________________________________

**Comment by AndyMcProducer** [comment]
Also, if it's using the system default set browser then any security software built in will still keep people protected from bad sites as much as if opening a browser window.
IF the chat can be limited html then the server screen could be limited, even if we have to make it and store it local in the software dir.
________________________________

**Comment by hoffie** [comment]
> But it should be possible to bring up a window that is a container of the default browser

Bringing up a new Window would essentially be supported already -- just place a link in the Welcome message and it should be opened in a default browser window.

I guess you are still refering to embedding web content into Jamulus. Qt supports such things. The component which uses the system browser (`WebView`) only seems to do that on mobile devices. On other platforms, it uses its own engine (`WebEngine`):

https://doc.qt.io/qt-5/qtwebengine-index.html
> This is useful on mobile platforms such as Android, iOS, and UWP (Universal Windows Platform); especially on iOS, where policy dictates that all web content is displayed using the operating system's web view.
> On Windows, Linux, and macOS, Qt WebView depends on the Qt WebEngine module to render content.

**Comment by gilgongo** [comment]
Hi @AndyMcProducer  I'm moving this to a Discussion until such time as we have a defined spec for what we want to do, whereupon we can raise it as a ticket to be worked on.
